### PR TITLE
Handle not compatible parameters to IndexOf

### LINKED ISSUE #84565: ImmutableArray cast to IList does not conform to other collection's usage
### Description

When casting an ImmutableArray to an IList, calls to IndexOf do not act as they do in other collections.  Other collections return -1 for items that are not found, regardless of their type.  ImmutableArray cast to IList then IndexOf is called will throw a NullReferenceException or InvalidCastException depending on the parameter passed in.
I originally found this when I noticed a Binding in WPF throwing an exception when bound to an ImmutableArray.  WPF was assuming that IndexOf on IList would always return -1 for a non-match.

### Reproduction Steps

Assuming the following Enum:
`enum States
{
    A,
    B
}`

The following code will show the problem:

    static void Main(string[] args)
    {
        List<States> list = new List<States>(new[] { States.A, States.B });
        TestList(list);

        States[] array = new[] { States.A, States.B };
        TestList(list);

        ImmutableArray<States> imm = ImmutableArray.Create(new[] { States.A, States.B });
        TestList(imm);
    }

    private static void TestList(IList list)
    {
        list.IndexOf(States.A);
        list.IndexOf(null);
        list.IndexOf(new object());
    }

TestList on ImmutableArray throws a NullReferenceException when ilist.IndexOf(null) is called.  If that line is commented out, it will throw an InvalidCastException when ilist.IndexOf(new object()) is called.

### Expected behavior

I would expect IList.IndexOf to return -1 and not throw one of those exceptions regardless of the implementing type.

### Actual behavior

A NullReferenceException or an InvalidCastException are thrown, depending on whether null is passed as a parameter, or an invalid type is passed as a parameter.

### Regression?

_No response_

### Known Workarounds

_No response_

### Configuration

Version:  7.0
OS:  Windows 10
Arch:  X64
I doubt it is unique to this OS or machine architecture.

### Other information

This may be because at https://github.com/dotnet/runtime/blob/a7f3df9b14d896cda708e2d22f35165737646769/src/libraries/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray_1.cs#L1211 the passed in parameter is directly cast to the type in the collection.

### PR DESCRIPTION
**Pull request by star-thoughts**
Fixes ImmutableArray<T> and SortedSet<T> for the following:
* IList.IndexOf is called with a null "value" parameter
* IList.IndexOf is called with an incorrectly-typed "value" parameter

Fix #84565

### CODE REVIEW

File: src/libraries/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray_1.cs [hunk 1]
```diff
@@ -1183,6 +1183,13 @@ void IList.Clear()
             throw new NotSupportedException();
         }
 
+        private static bool IsCompatibleObject(object? value)
+        {
+            // Non-null values are fine.  Only accept nulls if T is a class or Nullable<U>.
+            // Note that default(T) is not equal to null for value types except when T is Nullable<U>.
+            return (value is T) || (value == null && default(T) == null);
```
Comment (eiriktsarpalis) [inline src/libraries/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray_1.cs:1190]: `default(T) == null` is treated as a constant by the JIT so the RHS of the expression can be optimized away if you express it like this:
```suggestion
            return (value is T) || (default(T) == null && value == null);
```

File: src/libraries/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableSortedSet_1.cs [hunk 2]
```diff
@@ -840,6 +839,13 @@ void IList.Clear()
             throw new NotSupportedException();
         }
 
+        private static bool IsCompatibleObject(object? value)
+        {
+            // Non-null values are fine.  Only accept nulls if T is a class or Nullable<U>.
+            // Note that default(T) is not equal to null for value types except when T is Nullable<U>.
+            return (value is T) || (value == null && default(T) == null);
```
Comment (eiriktsarpalis) [inline src/libraries/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableSortedSet_1.cs:846]: See my feedback on the ImmutableArray method.

File: src/libraries/System.Collections.Immutable/tests/IndexOfTests.cs [hunk 3]
```diff
@@ -44,6 +44,8 @@ public static class IndexOfTests
             if (factory(list) is IList)
             {
                 Assert.Equal(-1, ((IList)factory(list)).IndexOf(6));
```
Comment (eiriktsarpalis) [inline src/libraries/System.Collections.Immutable/tests/IndexOfTests.cs:46]: This is testing `ImmutableList<T>`. While the additions are nice to have, could you also add a test for the updated `ImmutableArray` methods?

File: src/libraries/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray_1.cs (hunk 1 above)
Comment (eiriktsarpalis) [reply src/libraries/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray_1.cs:1190]: Sure, why not?

File: src/libraries/System.Collections.Immutable/tests/IndexOfTests.cs (hunk 3 above)
Comment (eiriktsarpalis) [reply src/libraries/System.Collections.Immutable/tests/IndexOfTests.cs:46]: Are these testing the IList IndexOf method and are they accounting for mistyped inputs?

Comment (star-thoughts) [reply src/libraries/System.Collections.Immutable/tests/IndexOfTests.cs:46]: If I understand your question correctly, yes. I put them in before making the fix and the test was failing until the fix went in.  They are testing for null and "a" on IList<int> collections.

Comment (eiriktsarpalis) [reply src/libraries/System.Collections.Immutable/tests/IndexOfTests.cs:46]: Ah I see, so presumably one of the `factory` parameters being passed by the test converts the `ImmutableList` to an `ImmutableArray`?

Comment (star-thoughts) [reply src/libraries/System.Collections.Immutable/tests/IndexOfTests.cs:46]: The factory parameter creates either an ImmutableArray or ImmutableList depending on what is being tested.

**Comment by eiriktsarpalis** [comment]
@star-thoughts when you get the chance, would it be possible to take a look at the remaining feedback? Thanks!

**Comment by star-thoughts** [comment]
Yes sorry.  I was out of town for the weekend, and got sick as soon as I returned.  Hopefully I'll have time to fix the rest of the items reported in the issue over the weekend.

**Review by eiriktsarpalis — APPROVED** [review]
Thanks!

**Pull request merged**